### ISSUE TITLE
Plugin causing ZBS not to start with no warning messages.

### ISSUE THREAD

**Issue by dbaileychess** [pkulchenko/ZeroBraneStudio#448]
I have a plugin I wrote in my packages/ directory called [tfs.lua](http://pastebin.com/tpRZLuEc). When it is present, ZBS will start up (according to Task Manager) but never show a window. Then after about 5 seconds will close (process removed from Task Manager). But it never gives me an error on why it doesn't start.

If I remove the tfs.lua file from the system, it boots up fine with no issues.

I have my cloned version of ZBS that is up-to-date with the upstream. I place that plugin in that directory tree and that version of ZBS starts up with no problem and the plugin is properly registered. 

I tried doing a git clean -f in your version of ZBS, and the only file it removes is the tfs.lua. And ZBS starts up fine after that. So it seems like the error handling for plugin is not working. But is it odd because the plugin works with no errors in my version of ZBS.

Is there any logs that I can look at to see why your version of ZBS is no loading?

**Comment by pkulchenko** [comment]
Derek, I don't see anything wrong with the plugin and it works on my copy. I also tried to make a compile error or a run-time error in the plugin (for example, replace `require` with `require`) and both generate a message in the Output window as expected.

It seems that this issue may be related to the fact that you are running external commands while the IDE is still being setup. I'd avoid any lengthy processing in the plugin body itself; all the setup code needs to be done in `onRegister`, but in this case I would do it as delayed initialization on first use.

> Is there any logs that I can look at to see why your version of ZBS is no loading?

When you say "your version", what do you mean? It appears that it's working fine with the up-to-date version from the master. What other version are you comparing it with?

When there are issues with ZBS code itself, you can run `bin\lua src\main.lua` to get the error reported to the standard output, so you may try that.

**Comment by pkulchenko** [comment]
I think I figured out what you mean by "my" and "your" version; it seems like you are comparing a local copy of your forked repository with a local copy of my repository. Since you indicated that your fork is up-to-date, both clones should be the same, no?

**Comment by dbaileychess** [comment]
I did the bin/lua.exe src/main.lua and it starts up fine, no errors. But using the zbstudio.exe still crashes though.

Yes, our two repos should be identical, so I don't know why it is working for one and not the other. I can try moving the external commands to the onRegister function and report back.

**Comment by pkulchenko** [comment]
> I can try moving the external commands to the onRegister function and report back.

Yes, I'd try that. I'd also try running it from the actual menu code (`if not self.tfexe then self.tfexe = ... end`).

I'm a bit at a loss what may be causing this issue you see, so any additional information may help. I don't see the issue, but it's no surprise as I don't have TFS installed.

**Comment by pkulchenko** [comment]
Derek, one more thing to try: instead of running the process through winapi, you can simply check for the file to exist using `wx.wxFileExists(file)`.

Also, ZBS provides facility for running external processes, which was designed specifically for things like this integration you are working on (I plan to have something similar with git/svn). It's not well documented, but it's quite simple (the code is in the `package.lua`). For example, try running this command from the local Console: 

```lua
ide:ExecuteCommand("git --no-pager blame main.lua", "",
  function(s) DisplayShell(s) end,
  function() DisplayShell("done") end)
```

Assuming ZBS/src is the current folder it should show "blame" result for `main.lua`.

**Comment by madmaxoft** [comment]
If it works in one repo and doesn't work in another, consistently, then it's a good way to find out what's wrong. Did you do a file-by-file contents comparison of the two repos? Maybe there is some difference after all, and the difference would be the cause for the bad behavior.

**Comment by dbaileychess** [comment]
Hi Mattes and Paul,

  Good idea, I did a file-by-file comparison and every file (other than
those in the .git directory) are identical between the two. And I am still
having the problem.

  Paul, I agree with all your suggestions on how to improve the plugin, but
it is still odd that the plugin works in my forked-repo and not your repo.
I moved all the init stuff into the onRegister() like you suggested. It
still works on my forked-repo, but the same file in your repo still fails
to load the IDE.

  I don't think that wx.wxFileExists(file) works with binary files,
according to the docs it is true if the file exists and is a plain file
<http://docs.wxwidgets.org/3.0/group__group__funcmacro__file.html#ga72d2e4ef03d51c0d59c2cb788462d600>.
It was always returning false for me, even though the file exists.

  I did a clean clone of your repo in a local directory called ZBSTest. I
started zbstudio.exe and it worked fine. I moved my plugin file to the
packages/ directory and the same issue appeared, it just wouldn't start.

  I did notice that if I moved the line: local G = ... from the first line
of my file into the onRegister() function, the IDE would start with the
following error:

"Error while loading file: 'packages\tfs.lua:10' cannot use '...' outside a
vararg function near '...''.

I also got a

 "The single instance communication has failed; there may be another
instance running, which may cause interference with the debugger".

But according to task manager, I only have one zbstudio.exe running (the
current one).

 I am still at a lost, here is my current plugin file
http://pastebin.com/4hihUHQb



Take Care,

    Derek

**Comment by pkulchenko** [comment]
Derek,

I'm certain that `wx.wxFileExists` works with executable files; I think by "plain" files they mean that it's not a device, a named pipe, or something similar. Try `wx.wxFileExists([[bin/lua.exe]])` from the console when the project is set to ZBS directory. Note that ZBS changes the current directory to the project directory when it's running commands in the Local Console (otherwise `require 'mymodule'` wouldn't work as expected if "mymodule" is in your project directory).

> I did notice that if I moved the line: local G = ... from the first line of my file into the onRegister() function, the IDE would start with the following error:

This is expected as ... refers to the vararg that is not defined for the callbacks. You don't need to use any `G` references in your callbacks as they are always executed in the right environment. This is only needed for the config or any initial plugin code.

> "The single instance communication has failed; there may be another instance running, which may cause interference with the debugger". But according to task manager, I only have one zbstudio.exe running (the current one).

When you get this error also check for the `lua` instance, not only for `zbstudio` instance.

> I am still at a lost

Unfortunately, I'm at a loss as well. I realize that switching to `wxFileExists` is not a fix, but it would at least provide us with some additional information and a workaround for you.

It looks like it works with `bin\lua` and doesn't work with `zbstudio` for you. The code for zbstudio launcher is in `build\win32_starter.c` and all it does is loads lua dll and loads and executes `src/main.lua`. I have a different started code in a separate branch that is waiting its time, but this code hasn't given me any trouble and has been used for a very long time on various platforms, so I don't have complaints about it.

All I can suggest at this point is to add `error(something)` code to the plugin, probably after `winapi.execute` call, to see if that point is reached or if the call is hanging somewhere. You can wrap it into an `if` if it's reached multiple times. If this doesn't help, I'd also add `error()` calls to `src\main.lua` to see where the executions stops or hangs.

In a similar cases in the past I used `log` function that looked like this: `function log(msg) local f = io.open("d:/log.out", "w") f:write(msg) f:close() end`. It's not fast, but simple and effective in gathering information. You may add `os.clock()` to include the timing.

One last option to consider is to set debug hook configured with function calls and to also get a log from it to track what is being executed (as the execution may be looping somewhere). I have a snippet that may be used for that.

**Comment by pkulchenko** [comment]
Derek, do you have any new information on this issue?

**Comment by pkulchenko** [comment]
Derek, please reopen if you have any new information on this issue.

**Comment by dbaileychess** [comment]
Hi Paul,

 Sorry about the lack of response, been busy at work and haven't looked
into it again until today.

I tried going to your suggestion with: wx.wxFileExists(path) but still this
never returns true even though that file exists. I figured out that I had
the path surrounded with " characters due to the spaces in the path. If I
remove those, i.e.

[[Path\to\the\file]] vs [["Path\to\the\file"]]

it works! This seems odd to me that quoting a path would cause it not to
work, but that is an issue with wx.

The reason I had the path quoted in the first place is that when I was
originally using 'winapi.execute(path)' to check if the program existed,
that function requires that the path be quoted if it contains internal
spaces. Otherwise it would give the error:

'"C:\Program' is not recognized as an internal or external command"

Since the path was 'C:\Program Files (x86)\Microsoft Visual Studio
12.0\Common7\Tools\..\IDE\TF.exe'

So in summary, I was passing the wrong path format into wx.wxFileExists.

Also, running the command winapi.execute(path) is what causes the ZBS not
to start up. I don't know what that is, but now that I can use
wx.wxFileExists I don't need to worry about that.

Take Care,

    Derek

**Comment by pkulchenko** [comment]
Derek, you are right; `wx.wxFileExists` take the path without any quotes (as it allows for spaces in file names), but any command that launches an application and takes parameters, would require quotes around filenames (to treat them as one entity, instead of multiple commands/parameters).

Glad the issue is resolved!
